src/transport/SecureSessionMgr.cpp

```diff
@@ -107,55 +133,98 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
     mPeerConnections.MarkConnectionActive(state);
 
     {
-        uint8_t * data = nullptr;
-        PacketHeader packetHeader;
+        uint8_t * data         = nullptr;
+        uint8_t * p            = nullptr;
+        uint32_t msgId         = 0;
+        uint32_t payloadLength = 0; // Make sure it's big enough to add two 16-bit ints without overflowing.
+        uint16_t len           = 0;
         MessageAuthenticationCode mac;
 
-        const uint16_t headerSize = payloadHeader.EncodeSizeBytes();
-        uint16_t actualEncodedHeaderSize;
-        uint16_t totalLen = 0;
-        uint16_t taglen   = 0;
-        uint32_t payloadLength; // Make sure it's big enough to add two 16-bit
-                                // ints without overflowing.
-        static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
-                      "Addition to generate payloadLength might overflow");
-        payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
-        VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
-
-        packetHeader
-            .SetSourceNodeId(mLocalNodeId)              //
-            .SetDestinationNodeId(peerNodeId)           //
-            .SetMessageId(state->GetSendMessageIndex()) //
-            .SetEncryptionKeyID(state->GetLocalKeyID()) //
-            .SetPayloadLength(static_cast<uint16_t>(payloadLength));
-        packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
+        if (!isEncrypted)
+        {
+            msgId = state->GetSendMessageIndex();
+
+            static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                          "Addition to generate payloadLength might overflow");
+
+            headerSize    = payloadHeader.EncodeSizeBytes();
+            payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+            VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+            packetHeader
+                .SetSourceNodeId(mLocalNodeId)              //
+                .SetDestinationNodeId(peerNodeId)           //
+                .SetMessageId(msgId)                        //
+                .SetEncryptionKeyID(state->GetLocalKeyID()) //
+                .SetPayloadLength(static_cast<uint16_t>(payloadLength));
+            packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
+        }
+        else
+        {
+            err = packetHeader.Decode(msgBuf->Start(), msgBuf->DataLength(), &headerSize);
```

Comment:
Isn't the packetHeader already decoded here: https://github.com/project-chip/connectedhomeip/pull/4048/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#diff-95eaad1efd04391d106529cad37a61fd4a010a561587a6ad5be004d00b22bfdeR115